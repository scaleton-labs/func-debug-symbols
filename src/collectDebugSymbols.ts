import { CompilerConfig } from '@ton-community/func-js';
import { DebugSymbols } from './types';
import { Cell } from '@ton/core';
import { createDebugCompiler } from './createDebugCompiler';
import { unpackDebugSymbols } from './unpackDebugSymbols';

export async function collectDebugSymbols(
  config: CompilerConfig,
): Promise<DebugSymbols> {
  const compiler = createDebugCompiler();
  const result = await compiler.compileFunc(config);

  if (result.status === 'error') {
    throw new Error('Cannot compile contacts: ' + result.message);
  }

  const rootCell = Cell.fromBase64(result.codeBoc);

  return unpackDebugSymbols(rootCell);
}
