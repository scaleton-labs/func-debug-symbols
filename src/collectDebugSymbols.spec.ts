import { CompilerConfig } from '@ton-community/func-js';
import { collectDebugSymbols } from './collectDebugSymbols';

describe('collectDebugSymbols', () => {
  it('collects information about procedures and globals', async () => {
    const config: CompilerConfig = {
      targets: ['main.fc'],
      sources: {
        'main.fc': `
          global int a;
          global cell b;
          global slice c;

          () throw_inline() impure inline { throw(1); }
          () throw_inline_ref() impure inline_ref { throw(2); }
          () throw_get() method_id(88) { throw(3); }
          () recv_internal() { throw(4); }
        `,
      },
    };

    const debugSymbols = await collectDebugSymbols(config);
    expect(debugSymbols).toMatchSnapshot();
  });
});
