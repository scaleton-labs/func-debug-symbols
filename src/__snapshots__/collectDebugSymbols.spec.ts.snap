// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collectDebugSymbols collects information about procedures and globals 1`] = `
{
  "globals": [
    {
      "index": 1,
      "name": "a",
    },
    {
      "index": 2,
      "name": "b",
    },
    {
      "index": 3,
      "name": "c",
    },
  ],
  "procedures": [
    {
      "cellHash": "b1b3726fb943fec8e144bb0e470a74a891fa4948fbc50b85ca1981e5444dfb4b",
      "methodId": 0,
      "name": "main",
    },
    {
      "cellHash": "f8664b544b6b2451eb1826d6fc9c70fcd94f89a092391fa68355ea1c08b51c0d",
      "methodId": 1,
      "name": "throw_inline",
    },
    {
      "cellHash": "3ef4fc929ecbb6c74979e7ec38a918e94c16496bb7c6ffa8a0905eb401b8f3be",
      "methodId": 2,
      "name": "throw_inline_ref",
    },
    {
      "cellHash": "588d70db264b8d53c52e563524749977fd62985d6bda14e3a9b83dd5ba9c9041",
      "methodId": 88,
      "name": "throw_get",
    },
  ],
}
`;
